File: cpp/MinHeap.cpp
#include "MinHeap.h"
#include <iostream>
#include <stdexcept>
#include <iomanip>

using namespace std;

// Restore min-heap property by bubbling element UP
// Called after insertion at the end of the array
void MinHeap::heapifyUp(int index) {
    while (index > 0 && heap[parent(index)].first > heap[index].first) {
        swap(heap[parent(index)], heap[index]);
        index = parent(index);
    }
}

// Restore min-heap property by sinking element DOWN
// Called after replacing root with last element during extractMin
void MinHeap::heapifyDown(int index) {
    int smallest = index;
    int left = leftChild(index);
    int right = rightChild(index);
    int n = (int)heap.size();

    if (left < n && heap[left].first < heap[smallest].first) {
        smallest = left;
    }
    if (right < n && heap[right].first < heap[smallest].first) {
        smallest = right;
    }

    if (smallest != index) {
        swap(heap[index], heap[smallest]);
        heapifyDown(smallest);
    }
}

// Insert a new (distance, point) pair into the heap
void MinHeap::insert(double distance, const Point& p) {
    heap.push_back(make_pair(distance, p));
    heapifyUp((int)heap.size() - 1);
}

// Remove and return the element with minimum distance
pair<double, Point> MinHeap::extractMin() {
    if (heap.empty()) {
        throw runtime_error("MinHeap::extractMin() called on empty heap");
    }

    pair<double, Point> minElement = heap[0];
    heap[0] = heap.back();
    heap.pop_back();

    if (!heap.empty()) {
        heapifyDown(0);
    }

    return minElement;
}

// View the minimum element without removing it
pair<double, Point> MinHeap::peek() const {
    if (heap.empty()) {
        throw runtime_error("MinHeap::peek() called on empty heap");
    }
    return heap[0];
}

// Display all elements in the heap (level-order)
void MinHeap::display() const {
    if (heap.empty()) {
        cout << "  (heap is empty)" << endl;
        return;
    }
    for (int i = 0; i < (int)heap.size(); i++) {
        cout << "  dist=" << fixed << setprecision(2) << heap[i].first
             << "  " << heap[i].second << endl;
    }
}

File: src/minheap.js
export class MinHeap {
    constructor() {
        this.heap = []; 
    }

    _parent(i)     { return Math.floor((i - 1) / 2); }
    _leftChild(i)  { return 2 * i + 1; }
    _rightChild(i) { return 2 * i + 2; }

    _swap(i, j) {
        [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
    }

    _heapifyUp(index) {
        while (index > 0 &&
               this.heap[this._parent(index)].distance > this.heap[index].distance) {
            this._swap(this._parent(index), index);
            index = this._parent(index);
        }
    }

    _heapifyDown(index) {
        let smallest = index;
        const left   = this._leftChild(index);
        const right  = this._rightChild(index);
        const n      = this.heap.length;

        if (left  < n && this.heap[left].distance  < this.heap[smallest].distance) smallest = left;
        if (right < n && this.heap[right].distance < this.heap[smallest].distance) smallest = right;

        if (smallest !== index) {
            this._swap(index, smallest);
            this._heapifyDown(smallest);
        }
    }

    insert(distance, point) {
        this.heap.push({ distance, point });
        this._heapifyUp(this.heap.length - 1);
    }

    extractMin() {
        if (this.heap.length === 0) throw new Error('MinHeap is empty');
        const min = this.heap[0];
        this.heap[0] = this.heap[this.heap.length - 1];
        this.heap.pop();
        if (this.heap.length > 0) this._heapifyDown(0);
        return min;
    }

    peek() {
        if (this.heap.length === 0) throw new Error('MinHeap is empty');
        return this.heap[0];
    }

    get size()    { return this.heap.length; }
    get empty()   { return this.heap.length === 0; }
}